<?xml version="1.0" encoding="UTF-8"?>

<mule
	xmlns:http="http://www.mulesoft.org/schema/mule/http"
	xmlns:cloudhub="http://www.mulesoft.org/schema/mule/cloudhub"
	xmlns:ee="http://www.mulesoft.org/schema/mule/ee/core"
	xmlns:module-error-handler-plugin="http://www.mulesoft.org/schema/mule/module-error-handler-plugin"
	xmlns:json-logger="http://www.mulesoft.org/schema/mule/json-logger"
	xmlns="http://www.mulesoft.org/schema/mule/core" xmlns:doc="http://www.mulesoft.org/schema/mule/documentation" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/json-logger http://www.mulesoft.org/schema/mule/json-logger/current/mule-json-logger.xsd
http://www.mulesoft.org/schema/mule/module-error-handler-plugin http://www.mulesoft.org/schema/mule/module-error-handler-plugin/current/mule-module-error-handler-plugin.xsd
http://www.mulesoft.org/schema/mule/ee/core http://www.mulesoft.org/schema/mule/ee/core/current/mule-ee.xsd
http://www.mulesoft.org/schema/mule/cloudhub http://www.mulesoft.org/schema/mule/cloudhub/current/mule-cloudhub.xsd
http://www.mulesoft.org/schema/mule/http http://www.mulesoft.org/schema/mule/http/current/mule-http.xsd">
	
	
	<error-handler name="api-error-handler">
		<on-error-propagate enableNotifications="true"
			logException="true" doc:name="On Error Propagate"
			doc:id="3d0958dd-ce68-43bd-94e1-a699d4b3c62f">
			<json-logger:logger doc:name="Log Exception" 
				config-ref="jsonLoggerConfig"
				doc:id="e39fe198-d7aa-401e-90e0-cb2865ea3228" priority="ERROR"
				tracePoint="EXCEPTION" message="An error has occured"/>
			<module-error-handler-plugin:process-error
				doc:name="Process Error"
				doc:id="1d795cc2-218e-48cc-b749-411a3c4c05a5"/>
			<set-variable value="#[attributes.httpStatus]"
				doc:name="Set Status" doc:id="a012e0ea-984f-404e-a87e-dd459df02f9c"
				variableName="httpStatus"/>
			<flow-ref doc:name="main-error-handler-flow" doc:id="04cd50c8-9e7c-47e1-a005-4a071fa315cf" name="main-error-handler-flow" />
			<json-logger:logger
				doc:name="Log End with Exception"
				doc:id="b1ef7bd4-7cf4-47b2-8fe4-bcbc08a29183" priority="ERROR"
				tracePoint="END" message="Finished with Error"
				config-ref="jsonLoggerConfig"/>
		</on-error-propagate>
	</error-handler>
	<sub-flow name="main-error-handler-flow" doc:id="ca8479da-4b7f-4e95-9310-cbc8897ffbab" >
		<ee:transform doc:name="hhtp status and error payload" doc:id="a4afdea9-810d-4a57-bb53-e6094bbbf59e">
				<ee:message>
					<ee:set-payload><![CDATA[%dw 2.0
output application/json
---

{
    "error": {
        "code": vars.httpStatus,
        "reason": error.errorType.identifier default 'APP:UNEXPECTED_ERROR',
        "message": error.description default 'Not handled error',
        "transaction": correlationId 
	}
}]]></ee:set-payload>
				</ee:message>
				<ee:variables>
					<ee:set-variable variableName="httpStatus"><![CDATA[%dw 2.0
output application/java
---
'400']]></ee:set-variable>
				</ee:variables>
			</ee:transform>
		<try doc:name="Try" doc:id="809b4e0a-0e86-46d8-8c56-3099623debc2">
				<cloudhub:create-notification domain="#[p('error.domain') ++ '-' ++ p('env')]" doc:name="Create Notification" doc:id="e225013a-86fb-46dc-9aa5-bb1832c5599b" config-ref="CloudHub_Config_OAuth_Client_Credentials" priority="ERROR">
					<cloudhub:message><![CDATA[#["Error Message:  " ++ payload.error.message]]]></cloudhub:message>
					<cloudhub:custom-properties><![CDATA[#[output application/java
---
{
	"code": payload.error.code,
    "reason": payload.error.reason,
    "transaction": payload.error.transaction
}]]]></cloudhub:custom-properties>
				</cloudhub:create-notification>
				<flow-ref doc:name="Log error in SF" doc:id="a61e5b89-9004-4c5a-adcb-1cac3e7345d7" name="log-error-sf" />
				<error-handler>
					<on-error-continue enableNotifications="true" logException="true" doc:name="On Error Continue" doc:id="f78abba4-4e76-4ac9-9ec3-10aadd580c52">
						<json-logger:logger doc:name="Error trying to raise a CH alert or loggin error in SF" doc:id="c464556a-a737-4961-ab04-3121ac60efbb" config-ref="jsonLoggerConfig" message="Error trying to raise a CH alert or loggin error in SF" tracePoint="EXCEPTION" priority="ERROR">
							<json-logger:content><![CDATA[#[import modules::JSONLoggerModule output application/json ---
{
    error: error
}]]]></json-logger:content>
						</json-logger:logger>
						<ee:transform doc:name="Add error logging flag" doc:id="5efea687-2030-419a-9ef4-0008b702bfea">
							<ee:message>
								<ee:set-payload><![CDATA[%dw 2.0
output application/json
---
payload ++ {"errorLogging": true}]]></ee:set-payload>
							</ee:message>
						</ee:transform>
					</on-error-continue>
				</error-handler>
			</try>
	</sub-flow>
	<sub-flow name="log-error-sf" doc:id="e4ffb435-0b83-4668-9c38-d31dbc8deb4f" >
		<ee:transform doc:name="To get log id" doc:id="b9293354-a107-4153-819a-1bfd9dab8171" >
			<ee:message />
			<ee:variables >
				<ee:set-variable variableName="logInSF" ><![CDATA[%dw 2.0
output application/json
---
{
  "transactionId": correlationId
}]]></ee:set-variable>
			</ee:variables>
		</ee:transform>
		<http:request method="POST" doc:name="POST /log  SFcore-sapi" doc:id="1fa9a1e9-e940-43c8-af25-51bbc802e83a" config-ref="HTTP_Request_sfCoreSapi" path="/log" target="logResponse" >
			<http:body ><![CDATA[#[vars.logInSF]]]></http:body>
			<http:headers ><![CDATA[#[output application/java
---
{
	"client_secret" : p('secure::error.sfcoresapi.client_secret'),
	"client_id" : p('secure::error.sfcoresapi.client_id')
}]]]></http:headers>
		</http:request>
		<ee:transform doc:name="To log entry" doc:id="d6319bf5-daee-46fe-b592-192291e4dee9" >
			<ee:message />
			<ee:variables >
				<ee:set-variable variableName="logEntryInSF" ><![CDATA[%dw 2.0
output application/json
---
{
  "transactionId": vars.logResponse.recordId,
  "message": error.description,
  "exceptionStackTrace": (error.cause.stackTrace map ((item, index) -> {stack: item.fileName ++ ' (line: ' ++ item.lineNumber ++')'})).stack joinBy  (', '),
  "exceptionMessage": error.cause.message
}]]></ee:set-variable>
			</ee:variables>
		</ee:transform>
		<http:request method="POST" doc:name="POST /log-entry  SFcore-sapi" doc:id="c7aafd78-90fa-4e02-9f68-a42c077921c8" config-ref="HTTP_Request_sfCoreSapi" path="/log-entry" target="logResponse" >
			<http:body ><![CDATA[#[vars.logEntryInSF]]]></http:body>
			<http:headers ><![CDATA[#[output application/java
---
{
	"client_secret" : p('secure::error.sfcoresapi.client_secret'),
	"client_id" : p('secure::error.sfcoresapi.client_id')
}]]]></http:headers>
		</http:request>
	</sub-flow>

</mule>
